ENH: allow categoricals in msgpack

pwaller opened the pull request:
Supercedes #12191.

I've made a best-effort to rebase this against master.

It seems to work here.

jreback commented:
- add a note in whatsnew/0.18.1

chris-b1 commented on pandas/io/packers.py:
```diff
@@ -506,11 +526,17 @@ def decode(obj):
+        ctor_dtype = np_dtype
+        if is_categorical_dtype(pd_dtype):
+            # Series ctor doesn't take dtype with categorical
+            ctor_dtype = None
```
@jreback  - should the `Series` constructor accept a `Categorical` with `dtype=category`?  This part felt hacky.

jreback replied:
it already does.
However in this case want to construct a ``pd.Categorical`` as you can pass all of the attributes thru
```
In [1]: Series(list('aabc'),dtype='category')
Out[1]: 
0    a
1    a
2    b
3    c
dtype: category
Categories (3, object): [a, b, c]
```

jreback replied:
but you don't need this here at all. the ``dtype`` is not present so its ``None``, the ``Series`` will just take a ``Categorical`` directly.

pwaller replied:
Can I conclude that no change is required here in scope of this PR?

jreback replied:
no, this needs to be reworked a bit, seem my 2nd comment.

chris-b1 replied:
@jreback - at this point `obj['data']` is already a `Categorical` (and dtype is category), which is why this was necessay?

```python
In [37]: pd.Series(pd.Categorical([1,2,3]), dtype='category')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-37-87a1a3228158> in <module>()
----> 1 pd.Series(pd.Categorical([1,2,3]), dtype='category')

ValueError: cannot specify a dtype with a Categorical
```

jreback replied:
hmm, that might be a bug as in this case it is *ok* to specify the dtype (but it should be ignored). can you open a separate issue for that.

pwaller replied:
I need to start a new thread on this line of code.

So, if I understood, it shouldn't be necessary to special case the `dtype`? If I remove the `dtype=ctor_dtype` change (and leave it set to `dtype=np_dtype`), then it does not work. I get:

```
ValueError: cannot specify a dtype with a Categorical unless dtype='category'
```

I don't currently understand what is being asked for here.

jreback replied:
you need to rebase on master, that's the issue that was just merged

pwaller replied:
This was after rebasing.

jreback replied:
so that's what I was first going to write. This is not a numpy-dtype and cannot be treated like one. Instead put this code in ``encode/decode`` (e.g. check ``isinstance(obj, Categorical)``). Then you won't have this issue at all.

pwaller commented:
> add a note in whatsnew/0.18.1

Done. Please let me know if it is to taste.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -23,6 +23,9 @@ Enhancements
 
 
+- ``read_msgpack`` now supports categoricals (:issue:`12573`)
+
```
support for serializing and de-serializing categoricalls with msgpack

pwaller replied:
Done.

jreback commented on pandas/io/packers.py:
```diff
@@ -214,6 +217,14 @@ def convert(values):
+                      'data': convert(values.codes)},
+            'categories': {'dtype': values.categories.dtype.name,
+                           'data': convert(values.categories.values)}
+        }
```
also need the ``ordered`` boolean (and a tests for it)

jreback commented on pandas/io/tests/test_packers.py:
```diff
@@ -333,7 +333,8 @@ def setUp(self):
                  [Timestamp('20130603', tz='CET')] * 3,
-            'G': [Timestamp('20130102', tz='US/Eastern')] * 5
+            'G': [Timestamp('20130102', tz='US/Eastern')] * 5,
+            'H': Categorical(['a', 'b', 'c', 'd', 'e'])
```
add in another ``Categorical`` with an ``ordered=True``. Further add one with datetimes as the categories

pwaller replied:
I added `'I': Categorical(['a', 'b', 'c', 'd', 'e'], ordered=True)` and it passes without any implementation, so I guess `assert_series_equal` is not sufficient? Should I create a new separate test which does the encode/decode and checks that the `ordered` value is correctly preserved?

pwaller commented on pandas/io/tests/test_packers.py:
```diff
@@ -341,6 +342,7 @@ def setUp(self):
         self.d['mixed'] = Series(data['E'])
         self.d['dt_tz_mixed'] = Series(data['F'])
         self.d['dt_tz'] = Series(data['G'])
+        self.d['categorical'] = Series(data['H'])
```
@chris-b1 is this correct? Why is it `Series` and not `Categorical`?

chris-b1 replied:
It wouldn't make a difference here, e.g. `df['c'] = Categorical(..)` and `df['c'] = Series(Categorical(..))` do the same thing

jreback commented:
ok #12574  has been merged. pls rebase/update.

pwaller commented:
Rebased.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -30,7 +30,7 @@ Enhancements
 
-
+- support for serializing and de-serializing categoricalls with msgpack (:issue:`12573`)
 
```
Support.........categoricals (sp)

jreback commented on pandas/io/packers.py:
```diff
@@ -214,6 +217,14 @@ def convert(values):
         return v.tolist()
 
+    if is_categorical_dtype(dtype):
+        return {
```
this can effectively be moved to ``encode`` with a type of ``Categorical``, rather than here

jreback replied:
actually leave this here (as its a dtype operation), but ALSO add ``Categorical`` directly to encode/decode. Need to add tests for that as well (e.g. just encoding/decoding a ``Categorical`` w/o a Series)

jreback replied:
so ignore my first comment here. move this logic to ``encode/decode``. then all will work.

jreback commented on pandas/io/packers.py:
```diff
@@ -247,6 +258,15 @@ def unconvert(values, dtype, compress=None):
     if as_is_ext:
         values = values.data
 
+    if is_categorical_dtype(dtype):
```
same here, should be in ``decode``

pwaller replied:
Done.

jreback commented on pandas/io/packers.py:
```diff
@@ -506,11 +526,17 @@ def decode(obj):
         np_dtype = pandas_dtype(dtype).base
+
+        ctor_dtype = np_dtype
+        if is_categorical_dtype(pd_dtype):
```
we shouln't need to do this gymnastics

pwaller replied:
Done.

pwaller commented:
Wow, I think I managed to clear all of your comments. Everything suddenly made sense in the end. Please let me know if I missed something and if I should squash or reword my commits.

pwaller commented:
I broke the NDFrame tests. Investigating.

jreback commented on pandas/io/packers.py:
```diff
@@ -319,6 +322,18 @@ def encode(obj):
+        return {u'typ': u'categorical',
+                u'klass': u(obj.__class__.__name__),
+                u'name': getattr(obj, 'name', None),
+                u'codes_dtype': u(obj._codes.dtype.name),
```
I don't think you actually need to specify the ``codes/catgories`` dtypes. Those will be automatic. IOW, don't use ``convert`` here, just same ``codes : obj.codes`` and ``categories : obj.categories``. These are recursive encoder/decoders.

pwaller replied:
Fixed.

jreback commented on pandas/io/packers.py:
```diff
@@ -502,10 +517,27 @@ def decode(obj):
+    elif typ == u'categorical':
+
+        codes = unconvert(obj[u'codes'],
+                          dtype_for(obj[u'codes_dtype']),
```
do you also don't need all this, just the ``.from_codes`` line

pwaller replied:
Do I misunderstand something?

When I do that:

```python

    elif typ == u'categorical':
        from_codes = globals()[obj[u'klass']].from_codes
        return from_codes(codes=obj['codes'],
                          categories=obj['categories'],
                          ordered=obj[u'ordered'],
                          name=obj[u'name'])
```

I get:

```pytb
Traceback (most recent call last):
  File "/home/pwaller/.local/src/pandas/pandas/core/categorical.py", line 412, in from_codes
    codes = np.asarray(codes, np.int64)
  File "/home/pwaller/.local/lib/python3.4/site-packages/numpy/core/numeric.py", line 474, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: invalid literal for int() with base 10: b'\x00\x01\x02\x03\x04'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pwaller/.local/src/pandas/pandas/io/tests/test_packers.py", line 374, in test_basic
    i_rec = self.encode_decode(i)
  File "/home/pwaller/.local/src/pandas/pandas/io/tests/test_packers.py", line 73, in encode_decode
    return read_msgpack(p, **kwargs)
  File "/home/pwaller/.local/src/pandas/pandas/io/packers.py", line 152, in read_msgpack
    return read(fh)
  File "/home/pwaller/.local/src/pandas/pandas/io/packers.py", line 137, in read
    l = list(unpack(fh, encoding=encoding, **kwargs))
  File "_unpacker.pyx", line 459, in pandas.msgpack._unpacker.Unpacker.__next__ (pandas/msgpack/_unpacker.cpp:4912)
  File "_unpacker.pyx", line 390, in pandas.msgpack._unpacker.Unpacker._unpack (pandas/msgpack/_unpacker.cpp:3985)
  File "/home/pwaller/.local/src/pandas/pandas/io/packers.py", line 525, in decode
    name=obj[u'name'])
  File "/home/pwaller/.local/src/pandas/pandas/core/categorical.py", line 415, in from_codes
    "codes need to be convertible to an arrays of integers")
ValueError: codes need to be convertible to an arrays of integers
```

pwaller replied:
Ah wait, I didn't do the encoding side correctly, please wait.

jreback replied:
yep

pwaller replied:
Please ignore. Fixed.

jreback commented on pandas/io/tests/test_packers.py:
```diff
@@ -363,7 +387,8 @@ def setUp(self):
             'F': [Timestamp('20130102', tz='US/Eastern')] * 5,
-            'G': [Timestamp('20130603', tz='CET')] * 5
+            'G': [Timestamp('20130603', tz='CET')] * 5,
+            'H': Categorical(['a', 'b', 'c', 'd', 'e'])
```
add with an ``ordered=True`` (as well)

pwaller replied:
Done.

jreback commented on pandas/io/packers.py:
```diff
@@ -319,6 +339,16 @@ def encode(obj):
+        return {u'typ': u'categorical',
+                u'klass': u(obj.__class__.__name__),
+                u'name': getattr(obj, 'name', None),
+                u'codes': obj._codes,
```
prob use ``obj.codes`` here (same but ``_codes`` is internal)

pwaller replied:
Done.

pwaller commented on pandas/io/tests/test_packers.py:
```diff
@@ -363,7 +387,9 @@ def setUp(self):
-            'G': [Timestamp('20130603', tz='CET')] * 5
+            'G': [Timestamp('20130603', tz='CET')] * 5,
+            'H': Categorical(['a', 'b', 'c', 'd', 'e']),
+            'I': Categorical(['a', 'b', 'c', 'd', 'e'], ordered=True),
```
I'm in a strange position here. I added this test - it passes, but I didn't add the relevant code in convert/unconvert to pass the ordered parameter through. What gives?

jreback replied:
that check is not used at all! I was just going to tell you to take it out. a ``Categorical`` is fully serialized/deserialized via ``encode/decode`` the ``dtype`` is NEVER ``category``. except when its a ``series`` but that is already handled.

jreback commented:
@pwaller can you rebase/update

jreback commented:
this looks pretty good code wise. can you squash everything.

jreback commented:
cc @kawochen

pwaller committed:
ENH: allow categoricals in msgpack

DOC: support for categoricals in read_msgpack

Add TestCategorical test cases

Add Catecorical ordered=True ndframe test

pwaller commented:
Squashed.

jreback commented on pandas/io/packers.py:
```diff
@@ -262,6 +265,9 @@ def convert(values):
+    if is_categorical_dtype(values):
+        return v
+
     if dtype == np.object_:
```
change line 271 to:

``elif is_object_dtype(dtype)``

pwaller replied:
Done

jreback commented on pandas/io/packers.py:
```diff
@@ -298,6 +304,9 @@ def unconvert(values, dtype, compress=None):
+    if is_categorical_dtype(dtype):
+        return values
+
     if dtype == np.object_:
```
same here

pwaller replied:
Done

jreback commented on pandas/io/packers.py:
```diff
@@ -393,6 +402,16 @@ def encode(obj):
+    elif isinstance(obj, Categorical):
+        return {u'typ': u'categorical',
+                u'klass': u(obj.__class__.__name__),
+                u'name': getattr(obj, 'name', None),
```
name was deprecated a while back, no need for it here

pwaller replied:
Done

jreback commented on pandas/io/packers.py:
```diff
@@ -576,10 +595,18 @@ def decode(obj):
+        return from_codes(codes=obj[u'codes'],
+                          categories=obj[u'categories'],
+                          ordered=obj[u'ordered'],
+                          name=obj[u'name'])
```
remove name

pwaller replied:
Done

pwaller commented:
Rebased and comments addressed.

pwaller commented:
I'm getting a test failure I'm not sure how to address and I don't have much time to look into it:

```
======================================================================
ERROR: test_basic_frame (pandas.io.tests.test_packers.TestNDFrame)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/io/tests/test_packers.py", line 410, in test_basic_frame
    i_rec = self.encode_decode(i)
  File "/home/travis/build/pydata/pandas/pandas/io/tests/test_packers.py", line 75, in encode_decode
    return read_msgpack(p, **kwargs)
  File "/home/travis/build/pydata/pandas/pandas/io/packers.py", line 203, in read_msgpack
    return read(fh)
  File "/home/travis/build/pydata/pandas/pandas/io/packers.py", line 188, in read
    l = list(unpack(fh, encoding=encoding, **kwargs))
  File "_unpacker.pyx", line 459, in pandas.msgpack._unpacker.Unpacker.__next__ (pandas/msgpack/_unpacker.cpp:4709)
  File "_unpacker.pyx", line 390, in pandas.msgpack._unpacker.Unpacker._unpack (pandas/msgpack/_unpacker.cpp:3843)
  File "/home/travis/build/pydata/pandas/pandas/io/packers.py", line 639, in decode
    blocks = [create_block(b) for b in obj[u'blocks']]
  File "/home/travis/build/pydata/pandas/pandas/io/packers.py", line 637, in create_block
    dtype=b[u'dtype'])
  File "/home/travis/build/pydata/pandas/pandas/core/internals.py", line 2518, in make_block
    return klass(values, ndim=ndim, fastpath=fastpath, placement=placement)
  File "/home/travis/build/pydata/pandas/pandas/core/internals.py", line 1903, in __init__
    placement=placement, **kwargs)
  File "/home/travis/build/pydata/pandas/pandas/core/internals.py", line 1314, in __init__
    raise TypeError("values must be {0}".format(self._holder.__name__))
TypeError: values must be Categorical
```

jreback commented on pandas/io/packers.py:
```diff
@@ -393,6 +402,16 @@ def encode(obj):
                     u'compress': compressor}
+
+    elif isinstance(obj, Categorical):
+        return {u'typ': u'categorical',
```
this needs to be 'catogory'

jreback commented on pandas/io/packers.py:
```diff
@@ -576,10 +595,18 @@ def decode(obj):
             result = result.tz_localize('UTC').tz_convert(tz)
         return result
 
+    elif typ == u'categorical':
```
'category'

pwaller commented:
Updated again. Thanks for your patience.

jreback commented:
To fix that error

```
diff --git a/pandas/io/packers.py b/pandas/io/packers.py
index ba38c8e..f009793 100644
--- a/pandas/io/packers.py
+++ b/pandas/io/packers.py
@@ -266,7 +266,7 @@ def convert(values):
 
     # convert object
     if is_categorical_dtype(values):
-        return v
+        return values
 
     if dtype == np.object_:
         return v.tolist()
```

pwaller commented:
Forced again.

jreback commented:
ok looks good. ping when green.

jreback commented:
thanks!
